src/Microsoft.AspNetCore.Routing/Matching/VectorizedRefEmitTrieJumpTable.cs

```diff
@@ -0,0 +1,548 @@
+                }
+
+                il.Emit(OpCodes.Br, next);
+
+                // Process the group
+                il.MarkLabel(inside);
+                EmitTable(il, group.ToArray(), index + 1, length, locals, labels, methods);
+                il.MarkLabel(next);
+            }
+
+            // goto: defaultDestination
+            il.Emit(OpCodes.Br, labels.ReturnDefault);
+        }
+
+        public static void EmitReturnDestination(ILGenerator il, (string text, int destination)[] entries)
+        {
+            Debug.Assert(entries.Length == 1, "We should have a single entry");
+            il.Emit(OpCodes.Ldc_I4, entries[0].destination);
+            il.Emit(OpCodes.Ret);
+        }
+
+        private static ulong GetUInt64Key(string text, int index)
+        {
+            Debug.Assert(index + 4 <= text.Length);
+            var span = text.ToLowerInvariant().AsSpan(index);
+            ref var p = ref Unsafe.As<char, byte>(ref MemoryMarshal.GetReference(span));
+            return Unsafe.ReadUnaligned<ulong>(ref p);
+        }
+
+        private static ushort GetUInt16Key(string text, int index)
+        {
+            Debug.Assert(index + 1 <= text.Length);
+            return (ushort)char.ToLowerInvariant(text[index]);
+        }
+
+        private class Locals
+        {
+            /// <summary>
+            /// Holds current character when processing a character at a time.
+            /// </summary>
+            public LocalBuilder UInt16Value { get; set; }
+
+            /// <summary>
+            /// Holds current character when processing 4 characters at a time.
+            /// </summary>
+            public LocalBuilder UInt64Value { get; set; }
+
+            /// <summary>
+            /// Used to covert casing. See comments where it's used.
+            /// </summary>
+            public LocalBuilder UInt64LowerIndicator { get; set; }
+
+            /// <summary>
+            /// Used to covert casing. See comments where it's used.
+            /// </summary>
+            public LocalBuilder UInt64UpperIndicator { get; set; }
+
+            /// <summary>
+            /// Holds a 'ref byte' reference to the current character (in bytes).
+            /// </summary>
+            public LocalBuilder P { get; set; }
+
+            /// <summary>
+            /// Holds the relevant portion of the path as a Span[byte].
+            /// </summary>
+            public LocalBuilder Span { get; set; }
+        }
+
+        private class Labels
+        {
+            /// <summary>
+            /// Label to goto that will return the default destination (not a match).
+            /// </summary>
+            public Label ReturnDefault { get; set; }
+
+            /// <summary>
+            /// Label to goto that will return a sentinel value for non-ascii text.
+            /// </summary>
+            public Label ReturnNotAscii { get; set; }
+        }
+
+        private class Methods
```

Comment:
The funny thing about the Unsafe class is that it only exists to do things that aren't expressible in C# but can be done in IL. You can skip generating these calls and instead use the IL directly:

https://github.com/dotnet/corefx/blob/e6506dbd5669b2996b3a1479e594e6e8a3a89774/src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il

Reply:
Good point. Though, the problem with using the IL directly is you could no longer translate the generated code back to C# to debug it.

Reply:
I think the effect is the same though no? I like that we can at least show the code that we're using in terms of c#.